app/models/account/master_methods.rb

```diff
@@ -8,7 +8,7 @@ module Account::MasterMethods
     alias_method :providers, :provider_accounts
 
     scope :by_provider_key, ->(provider_key) {
-      includes(:bought_cinstances).references(:bought_cinstances).merge(Cinstance.by_user_key(provider_key))
+      joins(:bought_cinstances).where(cinstances: { user_key: provider_key })
```

Comment:
This is not ideal because it would cause references not to be loaded which may lead to N+1 query issues.

See https://www.reddit.com/r/rails/comments/4ybybl/rails_5_activerecord_includes_joins_and/
>  * Includes is used if you want to use/display the associated records of what your are retrieving
>  * References is only used together with Includes when you use the associated table in a where or order clause (usually as a string)
>  * Joins is used when the associated table is needed in a Where or Order clause without preloading the associated records (they are not used/shown)

In case there is an order clause making troubles, perhaps try
```suggestion
        includes(:bought_cinstances).references(:bought_cinstances).merge(Cinstance.by_user_key(provider_key).reorder(''))
```

Reply:
So, at first I monkey-patched a method in oracle enhanced to make the original query work. But they I decided to try something different - specificaly, simplify the query so that it doesn't use `FIRST_VALUE`.

So, the violating query that was causing the `OCIError: ORA-01741: illegal zero-length identifier` error was:

```
SELECT * FROM (SELECT DISTINCT FIRST_VALUE("ACCOUNTS"."ID") OVER (PARTITION BY ["\"ACCOUNTS\".\"ID\""] ORDER BY "ACCOUNTS"."ID") AS alias_0__, "ACCOUNTS"."ID" FROM "ACCOUNTS" LEFT OUTER JOIN "CINSTANCES" ON "CINSTANCES"."USER_ACCOUNT_ID" = "ACCOUNTS"."ID" AND "CINSTANCES"."TYPE" = :a1 WHERE "CINSTANCES"."TYPE" = :a2 AND "CINSTANCES"."USER_KEY" = :a3 ORDER BY alias_0__ ) WHERE ROWNUM <= :a4  [["type", "Cinstance"], ["type", "Cinstance"], ["user_key", "[FILTERED]"], ["LIMIT", 1]]
```

Note the `["\"ACCOUNTS\".\"ID\""]` - this is wrong, and this is caused by this change supporting composite PK: https://github.com/rails/rails/commit/c18a95e38e9860953236aed94c1bfb877fa3be84

So, before a single column was returned, and now it's an array, and I am not sure if Oracle can support this somehow.

So, the original solution I tried (that worked by the way :sweat_smile: ) was monkey-patching the following method:

```

        # Fixing OCIError: ORA-01741: illegal zero-length identifier
        # because of https://github.com/rails/rails/commit/c18a95e38e9860953236aed94c1bfb877fa3be84
        # the value of `columns` is  [ "\"ACCOUNTS\".\"ID\"" ] which forms an incorrect query
        # ... OVER (PARTITION BY ["\"ACCOUNTS\".\"ID\""] ORDER BY "ACCOUNTS"."ID") ...
        def columns_for_distinct(columns, orders) # :nodoc:
          # construct a valid columns name for DISTINCT clause,
          # ie. one that includes the ORDER BY columns, using FIRST_VALUE such that
          # the inclusion of these columns doesn't invalidate the DISTINCT
          #
          # It does not construct DISTINCT clause. Just return column names for distinct.
          orders.reject(&:blank?).map { |s|
            s = visitor.compile(s) unless s.is_a?(String)
            # remove any ASC/DESC modifiers
            s.gsub(/\s+(ASC|DESC)\s*?/i, "")
          }.reject(&:blank?).map.with_index { |column, i|
            "FIRST_VALUE(#{column}) OVER (PARTITION BY #{columns.join(', ')} ORDER BY #{column}) AS alias_#{i}__"
          }
        end
```

The original implementation for reference: https://github.com/rsim/oracle-enhanced/blob/v7.1.1/lib/active_record/connection_adapters/oracle_enhanced_adapter.rb#L678-L692

So, the changes are:
1. `PARTITION BY #{columns}` => `PARTITION BY #{columns.join(', ')`, while I have no idea whether it would work with multiple columns :upside_down_face: 
2. `(order_columns << super).join(", ")` removed, because `super` was calling the original method, and produced the invalid query again.

This worked, but didn't seem quite right to me, so I decided to try simplifying the query instead, because honestly - I just don't quite understand how this query works :shrug: 

```
SELECT DISTINCT FIRST_VALUE("ACCOUNTS"."ID") OVER (PARTITION BY "ACCOUNTS"."ID" ORDER BY "ACCOUNTS"."ID")
```

Reply:
> This is not ideal because it would cause references not to be loaded which may lead to N+1 query issues.

@akostadinov what references?

I understand that this query searches for the Account only. From what I can tell, it doesn't care about the actual `bought_applications`, so doesn't need to preload them. And moreover, in theory there would only be one application always, as I don't think multiple accounts can have the same provider key. So, the query is "find me that single account whose bought application's user key is 'x'". So, I actually think a simple inner join with no preloading suits better in this case. But maybe I'm missing something... :thinking:

Reply:
I think originally they tried to reuse the scope for DRY purposes but then sifters should have been use for tihs purpose. 

On the other hand it is good to fix oracle to support what we previously had. Not in this PR though. I'll try to write a test for it.

Reply:
Thanks a lot for that investigation and your thoughts @akostadinov ! Very appreciated.

What is very interesting though is that your query does not have all these weird `SELECT DISTINCT FIRST_VALUE("ACCOUNTS"."ID") OVER (PARTITION BY ...`

And I've just realized that this is because you are using `.take`, and not `.first` as in the original code.

And in fact, just replacing `.first` with `.take` in the relevant methods avoids the issue - even with exactly the same scope. Given that we expect only a single account to match anyway, using `take` is perfectly fine, as it avoids the ordering. I think I'll go ahead and use `take` instead of `first`.

But as I had already changed the scope too, and you have verified that this change is OK, I will keep it also :)